metrics_test.go

```diff
@@ -0,0 +1,326 @@
+		"METRICS_PREFIX": "prefix1.",
+	}
+	tests.WithEnvVars(env, func() {
+		metrics.initMetricsSender()
+		assert.Len(t, metrics.Clients, 1)
+	})
+	notifyToClose(end)
+}
+
+/*
+Tests that if there is only METRICS_HOST (as well as PORT and PREFIX), there is only one client
+*/
+func TestInitDestinationClients_FailureErrorJSON(t *testing.T) {
+	resetMetricsSender()
+
+	env := map[string]string{
+		// Add several colons to create an unmarshalling error
+		"METRICS_DESTINATIONS": `[{"METRICS_HOST":::: "127.0.0.1","METRICS_PORT":"8125","METRICS_PREFIX": "testPrefix."}]`,
+	}
+
+	tests.WithEnvVars(env, func() {
+		assert.Panics(t, func() {
+			metrics.initDestinationClients(defaultClientBuilder)
+		})
+		assert.Len(t, metrics.Clients, 0)
+	})
+}
+
+/*
+Tests that if there is only METRICS_DESTINATIONS, with one element in the array, it builds one client
+*/
+func TestInitDestinationClients_SuccessOneDestination(t *testing.T) {
+	resetMetricsSender()
+
+	end := make(chan string)
+	port := getUniquePort()
+	go simpleUDPServer(localhost, port, end)
+	<-end
+
+	env := map[string]string{
+		"METRICS_DESTINATIONS": fmt.Sprintf(`[{"METRICS_HOST": "127.0.0.1","METRICS_PORT":"%s","METRICS_PREFIX": "testPrefix."}]`, port),
+	}
+
+	tests.WithEnvVars(env, func() {
+		metrics.initDestinationClients(defaultClientBuilder)
+		assert.Len(t, metrics.Clients, 1)
+	})
+	notifyToClose(end)
+}
+
+/*
+Tests that if there is only METRICS_DESTINATIONS, with twos elements in the array, it builds two clients
+*/
+func TestInitDestinationClients_SuccessMultipleDestinations(t *testing.T) {
+	resetMetricsSender()
+	end1 := make(chan string)
+	end2 := make(chan string)
+	port1 := getUniquePort()
+	port2 := getUniquePort()
+	go simpleUDPServer(localhost, port1, end1)
+	go simpleUDPServer(localhost, port2, end2)
+	<-end1
+	<-end2
+
+	env := map[string]string{
+		"METRICS_DESTINATIONS": fmt.Sprintf(`[{"METRICS_HOST": "127.0.0.1","METRICS_PORT":"%s","METRICS_PREFIX": "testPrefix."}, {"METRICS_HOST": "127.0.0.1","METRICS_PORT":"%s","METRICS_PREFIX": "testPrefix2."}]`, port1, port2),
+	}
+
+	tests.WithEnvVars(env, func() {
+		metrics.initDestinationClients(defaultClientBuilder)
+		assert.Len(t, metrics.Clients, 2)
+	})
+	notifyToClose(end1)
+	notifyToClose(end2)
+}
+
+/*
+Tests that it returns the metrics handler
+*/
+func TestGetMetricsSender(t *testing.T) {
+	resetMetricsSender()
+
+	end := make(chan string)
+	port := getUniquePort()
+	go simpleUDPServer(localhost, port, end)
+	<-end
+
+	env := map[string]string{
+		"METRICS_HOST":   localhost,
+		"METRICS_PORT":   port,
+		"METRICS_PREFIX": "prefix1.",
+	}
+
+	tests.WithEnvVars(env, func() {
+		metricsSender := GetMetricsSender()
+		assert.Len(t, metrics.Clients, 1)
+		assert.Equal(t, metrics, metricsSender)
+	})
+	notifyToClose(end)
+}
+
+/*
+Tests that an increment sends a metric
+*/
+func TestIncrement_Success(t *testing.T) {
+	resetMetricsSender()
+
+	port := getUniquePort()
+
+	pc, err := net.ListenPacket("udp", localhost+":"+port)
```

Comment:
I need the handler on pc, which simpleUDPServer does not return (considering its a goroutine) I would need to send the the bytes read through a channel (unless you got another idea). Seems like much pain and especially less readability